File: pset6/vigenere/vigenere.py
import sys
from cs50 import get_string
import string


def main():
    if len(sys.argv) == 2:
        k = sys.argv[1]
        if k.isalpha():
            plain_text = get_string("plaintext: ")
            s = cipher(plain_text,k)
            print(f"ciphertext:",end=" ")
            print(s)
        else:
            print("Usage: ./vigenere k")
            exit(1)
    else:
        print("Usage: ./vigenere k")
        exit(1)


def cipher(text, key):
    ciphered = ""
    n = len(text)
    k = len(key)
    offset = 0
    value = 0
    for i in range(n):
        if not((text[i] >= 'a' and text[i] <= 'z') or (text[i] >= 'A' and text[i] <= 'Z'))  :
            ciphered = ciphered + text[i]
        else:
            #print(f"{text[i]}", end=" ")
            char = key[offset]
            if char >= 'a' and char <= 'z':
                value = ord(char) - 97
            elif char >= 'A' and char <= 'Z':
                value = ord(char) - 65
            #print(f"{char}", end=" ")
            offset = offset + 1
            if offset == k:
                offset = 0
            if text[i] >= 'a' and text[i] <= 'z':
                #print(f"{value}", end=" ")
                new_char_value = ord(text[i]) + value
                #print(f"{new_char_value}", end=" ")
                if new_char_value > 122:
                    new_char_value = new_char_value - 26
                    #print(f"{new_char_value}", end=" ")
            elif text[i] >= 'A' and text[i] <= 'Z':
                #print(f"{value}", end=" ")
                new_char_value = ord(text[i]) + value
                #print(f"{new_char_value}", end=" ")
                if new_char_value > 90:
                    new_char_value = new_char_value - 26
                    #print(f"{new_char_value}", end=" ")
            new_char = chr(int(new_char_value))
            ciphered = ciphered + new_char
            #print(ciphered)
    return ciphered


if __name__ == "__main__":
    main()

File: pset2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

bool check1(int n);
bool check2(string k);
string chiper(string text,string key);
int change(char c);

int main (int argc, string argv[])
{
    // check if only two commands entered
    if (check1(argc))
        {
            string k = argv[1];
            // check if key consists of letters only
            if (check2(k))
                {
                    printf ("plaintext: ");
                    string plain_text = get_string();
                    // a Function to the chiper text
                    string s = chiper(plain_text,k);
                    printf ("ciphertext: ");
                    printf("%s",s);
                    // int n = strlen(s);
                    // char array[n+1];
                    // for (int i=0; i<n; i++)
                    //     array[i] = s[i];
                    // array[n+1] = '\0';
                    // printf("%s",array);

                }
            else
                {
                    printf ("Usage: ./vigenere k\n");
                    return 1;
                }
            printf("\n");
            return 0;
          }
    else
        {
            printf ("Usage: ./vigenere k\n");
            return 1;
        }
}

bool check1(int n)
{
    if (n == 2)
        {
            return true;
        }
    else
        {
            return false;
        }
}

bool check2(string k)
{
    for(int i = 0,j = strlen(k); i < j; i++ )
        {
            // check if k is only letters not numbers
            if (!((k[i] >= 'a' && k[i] <= 'z') || (k[i] >= 'A' && k[i] <= 'Z')))
                {
                    return false;
                }
        }

    return true;
}

// To change each letter to corresponding number
int change(char c)
{
    char small = 'a';
    char cap = 'A';
    int key = 0;
    for (int i =  0; i < 26; i++)
        {
            if (c == small || c == cap )
                {
                    return key;
                }
            small++;
            cap++;
            key++;
        }
    return key;
}

string chiper(string text,string key)
{
    int value = 0;
    int offset = 0;
    int key_digits = strlen(key);
    for (int i = 0,n = strlen(text);i < n;i++)
        {
            if (text[i] == ' ')
                {
                    continue;
                }
            else
                {
                    value = change(key[offset]);
                    if (text[i] <= 'z' && text[i] >= 'a')
                        {
                            if ((text[i] + value) <= 'z')
                                {
                                    text[i] = text[i] + value;
                                    offset++;
                                }
                            else
                                {
                                    text[i] = text[i] -26;
                                    text[i] = text[i] + value;
                                    offset++;
                                }
                        }
                    else if (text[i] <= 'Z' && text[i] >= 'A')
                        {
                            if ((text[i] + value) > 'Z')
                                {
                                    text[i] = text[i] -26;
                                    text[i] = text[i] + value;
                                    offset++;

                                }
                            else
                                {
                                    text[i] = text[i] + value;
                                    offset++;
                                }

                        }
                }
            if (offset == key_digits)
                {
                    offset = 0;
                }
        }
        return text;
}